nilearn/glm/first_level/first_level.py

```diff
@@ -122,25 +141,39 @@ def run_glm(Y, X, noise_model='ar1', bins=100, n_jobs=1, verbose=0):
     # Create the model
     ols_result = OLSModel(X).fit(Y)
 
-    if noise_model == 'ar1':
+    if 'ar' in noise_model:
+        ar_order = int(re.split('ar', noise_model)[1])
         # compute and discretize the AR1 coefs
-        ar1 = (
-            (ols_result.residuals[1:]
-             * ols_result.residuals[:-1]).sum(axis=0)
-            / (ols_result.residuals ** 2).sum(axis=0)
-        )
+        ar1 = [_yule_walker(ols_result.residuals[:, res].reshape(-1, 1).T,
+                            order=ar_order)
+               for res in range(ols_result.residuals.shape[1])]
+        if len(ar1[0]) == 1:
+            ar1 = np.asarray(ar1).ravel()
         del ols_result
-        ar1 = (ar1 * bins).astype(np.int) * 1. / bins
-        # Fit the AR model acccording to current AR(1) estimates
+        for idx in range(len(ar1)):
+            ar1[idx] = (ar1[idx] * bins).astype(np.int) * 1. / bins
```

Comment:
I'm not sure what this does.
For AR(1) it was simple to cluster voxels according to AR coef, but how should we do here ?

Reply:
Yeah I wasn't sure what the clustering was for, I assumed it was something MRI specific that I didn't know about. Can you explain what the clustering is for then we could come up with the best AR(N) approach. We could cluster by first coef. We could not cluster. We could cluster by residual.

Reply:
I don't understand what the clustering and labelling is for, I assumed it was something MRI specific. Could you please explain the purpose of the clustering, then maybe I can suggest an appropriate approach for AR(N) models. We could cluster by first coef, or by residual, or by sum of coef, or not cluster.

Reply:
Clustering is used to reduce computation: the regression fit, that includes whitening, is done on groups of voxels insead of isolated voxels: looping of a few tens of groups is much more efficient than looping over voxels. Bzw, we need to adapt the whitening scheme implemented in  `model.py` and `regression.py` modules.

Reply:
Great. Thank you for explaining that, it makes sense to me now. I will also take a look at `model.p` and `regression.py`.

So it seems we want to cluster based on something that will be uniformly distributed then. Could we use the last term in the AR fit then? I assume this term is the most likely to be noise, whereas I have observed the first fit term is often common across channels.

And apologies for differences we might have in terminology, I am not an MRI researcher.

Reply:
> So it seems we want to cluster based on something that will be uniformly distributed then. Could we use the last term in the AR fit then? I assume this term is the most likely to be noise, whereas I have observed the first fit term is often similar across channels.

Not sure I understand your suggestion: you want voxels in a given cluster to be as homogeneous as possible in terms of noise characteristics. So you would typically use all AR coefficients, but the most informative ones are the first and second coefficients.

Reply:
Sorry I previously misunderstood.

> So you would typically use all AR coefficients

Ok this is what the code is now doing. But please let me know if I am still mistaken.

In regards to your original comment about what the binning now does, I have added the following to the docs to inform the user that the binning is now per AR coefficient. Does this make sense? Or do you suggest we cluster differently?

```
    bins : int, optional
        Maximum number of discrete bins for the AR(1) coef histogram.
        If higher order AR models are specified this specifies the maximum
        number of bins per AR coefficient.
```